resteasy-core-spi/src/main/java/org/jboss/resteasy/api/validation/ResteasyViolationException.java

```diff
@@ -189,14 +190,16 @@ public String toString()
    {
       convertViolations();
       StringBuffer sb = new StringBuffer();
-      for (Iterator<List<ResteasyConstraintViolation>> it = violationLists.iterator(); it.hasNext();)
-      {
-         List<ResteasyConstraintViolation> violations = it.next();
-         for (Iterator<ResteasyConstraintViolation> it2 = violations.iterator(); it2.hasNext();)
-         {
-            sb.append(it2.next().toString()).append('\r');
+
+      CopyOnWriteArrayList<List<ResteasyConstraintViolation>> imutableViolations =
+              new CopyOnWriteArrayList<>(violationLists);
```

Comment:
As mentioned I think we need a larger change here. However this don't fix the issue as the `violationLists` is still iterated when creating this new list. We also don't really need a concurrent list in the `toString()` itself as once it's initialized it won't need to be thread-safe.

Reply:
Hi @jamezp, thank for sharing your thought.

At the first invocation, the list is created and the second invocation recreates the same list (with the same elements). I'm not avoiding the recreation here as I want to avoid a side effect applying a big change, but I think the better solution is to create this list at the constructor. Thus it does not need to be created on the toString method. This solution solves the java.util.ConcurrentModificationException. In my tests, it has worked fine. I think to solve the issue immediately we can use the  CopyOnWriteArrayList (thus the recreation of the list will not generate the issue) and a second time I can send a PR to solve the recreation. However, I think the solution about the recreation needs more tests. Let me know if it makes sense for you.

Reply:
Hi @jamezp,
I'm sorry, but I think I didn't explain well my thought. 

The method convertViolations() called by toString initiate the ArrayList. I think the final solution is to call this method inside the ResteasyConstraintViolation' constructor. I don't think the state of this ArrayList (violationList) should be changed after the ResteasyConstraintViolation construction. About the CopyOnWriteArrayList it solves the question about the java.util.ConcurrentModificationException, as at the iteration time it creates an immutable copy of the list. However, I think the final solution is to call the convertViolations() inside all ResteasyConstraintViolation' constructors. The main issue is the toString is updating the internal state. But, we need to check if it will generate a side effect. Thus, I think we can apply this solution and after apply the solution calling the convertViolations() inside the ResteasyConstraintViolation' constructors. What do you think about the final solution?

Reply:
Hi @rhuan080 I don't think that would quite work. The `convertViolations()` would initialize itself with empty lists. With the current design it needs to be invoked several times. It's one of the reason I think we should look at redesigning it. TBH i see other issues with it as well like `line = br.readLine()` executing twice which looks like it would skip a line with each read.

I've got some time to look at this unless you'd rather have a go at it.

Reply:
Hi @jamezp, thank you. Yes, I have checked you talked about the `convertViolations()`. About the `line = br.readLine()`, I think it is not a problem as it is a stream of stringRep. As it is a string it is thread-safe. I cannot see what is the problem here. If possible can you clarify for me? 

Before using the CopyOnWriteArrayList I have created a solution using `synchronized` and I was able to solve it, but I have applied updates in many methods. Do you want to suggest some solution like it?

Reply:
Hi @rhuan080 I just meant the double `br.readLine()` skips a line each time so something like:
```
line 1
line 2
line 3
line 4
```

All we'd see is line1 and line 3 with line 2 and line 4 being skipped.

If we're just trying to fix this one issue then even using an instance of a new `ArrayList`, e.g. swap your `CopyOnWriteArrayList` with `ArrayList`, would work. However the fact that we're seeing a `ConcurrentModificationException` sounds like we've got some currency issues. If we don't care about missing violations, then we can just do what you're doing, but use an `ArrayList` instead.

Reply:
Hi @jamezp, I agreed this class should be reviewed, however, I think it is better to open an issue and solve it in another PR. The PR was opened to just solve the ConcurrentModificationException. I agreed it should be reviewed, but we can open an issue discuss what is a better design for this class. Looking at this class I think it is overly complex. What do you think about the issue? Do you agree to open an issue to it and discuss it there?

Reply:
Yes I agree. We should probably do different fixes for different branches IMO. We don't want to touch the 3.x branches too much so leaving those with a fix like this might be okay. I'll need to look at the source a bit more, but we likely have potential issues exposed on the other methods as well.

I will say I don't think we need to use a `CopyOnWriteArrayList` for the local instance though. If we use it for the global instance that is okay though.

Reply:
Hi @jamezp, I'm sorry for my delay. I have opened this JIRA --> https://issues.redhat.com/browse/RESTEASY-2927
If possible, join this JIRA to start the discussion about the review.

Reply:
@rhuan080 this would be my suggestion for the minimal change https://github.com/resteasy/Resteasy/compare/main...jamezp:RESTEASY-2914.

Reply:
@jamezp I'll update the branches.

Reply:
Node!